pkg/cableengine/ipsec/strongswan.go

```diff
@@ -0,0 +1,397 @@
+package ipsec
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/bronze1man/goStrongswanVici"
+	"github.com/kelseyhightower/envconfig"
+	"github.com/submariner-io/submariner/pkg/types"
+	"k8s.io/klog"
+)
+
+const (
+	pidFile = "/var/run/charon.pid"
+
+	// DefaultReplayWindowSize specifies the replay window size for charon
+	DefaultReplayWindowSize = "1024"
+
+	// DefaultIkeSaRekeyInterval specifies the default rekey interval for IKE_SA
+	DefaultIkeSaRekeyInterval = "4h"
+
+	// DefaultChildSaRekeyInterval specifies the default rekey interval for CHILD_SA
+	DefaultChildSaRekeyInterval = "1h"
+)
+
+type strongSwan struct {
+	sync.Mutex
+
+	localSubnets              []string
+	localEndpoint             types.SubmarinerEndpoint
+	secretKey                 string
+	replayWindowSize          string
+	ipSecIkeSaRekeyInterval   string
+	ipSecChildSaRekeyInterval string
+
+	debug   bool
+	logFile string
+}
+
+type specification struct {
+	PSK     string
+	Debug   bool
+	LogFile string
+}
+
+func NewStrongSwan(localSubnets []string, localEndpoint types.SubmarinerEndpoint) (Driver, error) {
+	ipSecSpec := specification{}
+
+	err := envconfig.Process("ce_ipsec", &ipSecSpec)
+	if err != nil {
+		return nil, fmt.Errorf("error processing environment config for ce_ipsec: %v", err)
+	}
+
+	return &strongSwan{
+		replayWindowSize:          DefaultReplayWindowSize,
+		ipSecIkeSaRekeyInterval:   DefaultIkeSaRekeyInterval,
+		ipSecChildSaRekeyInterval: DefaultChildSaRekeyInterval,
+		localEndpoint:             localEndpoint,
+		localSubnets:              localSubnets,
+		secretKey:                 ipSecSpec.PSK,
+		debug:                     ipSecSpec.Debug,
+		logFile:                   ipSecSpec.LogFile,
+	}, nil
+}
+
+func (i *strongSwan) Init() error {
+	if err := runCharon(i.debug, i.logFile); err != nil {
+		return err
+	}
+
+	if err := i.loadConns(); err != nil {
+		return fmt.Errorf("Failed to load connections from charon: %v", err)
+	}
+	return nil
+}
+
+func (i *strongSwan) ConnectToEndpoint(endpoint types.SubmarinerEndpoint) (string, error) {
+	client, err := getClient()
+	if err != nil {
+		return "", err
+	}
+	defer client.Close()
+
+	return i.installCableInternal(endpoint, client)
+}
+
+func (i *strongSwan) installCableInternal(endpoint types.SubmarinerEndpoint, client *goStrongswanVici.ClientConn) (string, error) {
```

Comment:
This was named to previously correspond to `InstallCable` so really should be named `connectToEndpointInternal` however this function can just be inlined within `ConnectToEndpoint`.  Originally `installCableInternal` was extracted as it was called from another function that has since been removed.

Reply:
I think I meant to do that initially and it skipped my mind, inlined it.